### ISSUE TITLE
Auth Problem

### ISSUE THREAD

**Issue by epicbenz** [poblouin/homebridge-spotify-speaker#54]
**Describe Your Problem:**

Last Time, I promise :D

I have tried to set this plugin for a good friend of me (after it worked so well, thanks again). 

But with his Spotify Account (Client ID, Client Secret and AuthCode) I get the error down below. 
On his Amazon Echo is the same Spotify account (in Spotify Skill) from which is also the client ID, etc..

I think I did everything exactly like on my Homebridge, don't know where the mistake is.

Could it be, that this isn't working with a Turkish Spotify Account? Because this is the only difference I could tell to mine.  

I have tried to delete and create a new App in the developer's dashboard a few times but it still won't work with his account.

I also have tried to generate a new AuthCode by access to: 

https://accounts.spotify.com/authorize?client_id="client-ID"&redirect_uri=https://example.com/callback&scope=user-read-playback-state user-modify-playback-state user-read-currently-playing

**Logs:**

```
[8/10/2023, 9:59:19 PM] [HomebridgeSpotifySpeaker] Could not authorize Spotify:

 {"body":{"error":"invalid_client"},"headers":{"date":"Thu, 10 Aug 2023 19:59:19 GMT","content-type":"application/json","content-length":"26","set-cookie":["__Host-device_id=AQBQrPAYc5CG5kwPfyqpIfRQI5dupGRSjmOzHCrA52AUE4gbdQ0gfscv8RuV-RK34Ae66K4Pi6WDgvnxvfQdW5RD-doD3E_Z2TY;Version=1;Path=/;Max-Age=2147483647;Secure;HttpOnly;SameSite=Lax","sp_tr=false;Version=1;Domain=accounts.spotify.com;Path=/;Secure;SameSite=Lax"],"sp-trace-id":"881b6a11fa41f164","x-envoy-upstream-service-time":"6","server":"envoy","strict-transport-security":"max-age=31536000","x-content-type-options":"nosniff","via":"HTTP/2 edgeproxy, 1.1 google","alt-svc":"h3=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000","connection":"close"},"statusCode":400}
[8/10/2023, 9:59:19 PM] [HomebridgeSpotifySpeaker] We could not fetch the Spotify tokens nor authenticate using the code grant flow.
        Please redo the manual login step, provide the new auth code in the config then try again.
```

**Plugin Config:**

```

{
            "name": "HomebridgeSpotifySpeaker",
            "spotifyClientId": „ClientID",
            "spotifyClientSecret": „Client“Secret,
            "spotifyRedirectUri": "https://example.com/callback",
            "spotifyAuthCode": "AuthCode“,
            "spotifyPollInterval": 20,
            "deviceNotFoundRetry": {
                "enable": true,
                "retryDelay": 2000
            },
            "devices": [
                {
                    "deviceName": "Test",
                    "deviceType": "speaker",
                    "spotifyDeviceName": "xxx' 2. Echo Dot",
                    "spotifyPlaylistUrl": "https://open.spotify.com/playlist/xxxxxxPV2ffFrT8yX6sYCT7A_aw",
                    "playlistShuffle": true,
                    "playlistRepeat": false
                }
            ],
            "debug": false,
            "platform": "HomebridgeSpotifySpeaker"
        }
```

**Screenshots:**

**Environment:**

- **Plugin Version**: v.1.3.1
- **Homebridge Version**: v.1.6.1
- **Node.js Version**: v18.15.0
- **NPM Version**: 
- **Operating System**: Synology NAS

**Comment by joeyhage** [comment]
@epicbenz Turkish shouldn’t be an issue. Since you are able to retrieve an AuthCode, it should work.

Double check these:

```
"spotifyClientId": "abcd1234",
"spotifyClientSecret": "abcd1234",
…
"spotifyAuthCode": "abcd1234-1234_abcd",
```

spotifyClientId and spotifyClientSecret should have no special characters, only a-z, A-Z, 0-9.

spotifyAuthCode is the same but can also have `_` and `-`. 

Make sure the quotes `"` are good, too. Only vertical

I noticed from your question some of the quotation marks are curly/angled. The mark should be two vertical lines.  

It’s best to use the [homebridge ui plug-in](https://www.npmjs.com/package/homebridge-config-ui-x) and configure that way if you aren’t instead of editing the JSON file.

**Comment by epicbenz** [comment]
> @epicbenz Turkish shouldn’t be an issue. Since you are able to retrieve an AuthCode, it should work.
>
> I noticed from your question some of the quotation marks are curly/angled. The mark should be two vertical lines.  
> 
> 
> 
> It’s best to use the [homebridge ui plug-in](https://www.npmjs.com/package/homebridge-config-ui-x) and configure that way if you aren’t instead of editing the JSON file.


Ah it's because of the autocorrection of my mac, my bad. 
I used the Homebridge ui to config instead of editing the JSON file and copied step by step, but i will try it again. 

Can it be that anything isnt correct configurated on the Echo Dot itself? Actually i only have to Login on the Spotify Skill on the Echo with the same account from which the Client-ID is?

**Comment by joeyhage** [comment]
That error is happening before the connection to the echo dot - it is unable to login to Spotify at all. 

But yes, as long as the Spotify skill account matches the Client ID (Spotify developer account), you’re all set!

**Comment by epicbenz** [comment]
> That error is happening before the connection to the echo dot - it is unable to login to Spotify at all. 
> 
> 
> 
> But yes, as long as the Spotify skill account matches the Client ID (Spotify developer account), you’re all set!


Hmm really dont know where the error could be. If i enter my ID etc. it works, but if i enter his it wont work.

**Comment by joeyhage** [comment]
Try this url for the Auth Code and use his Client ID. In the url you provided above, you were missing a key part of the url:

https://accounts.spotify.com/authorize?client_id=abcd1234&response_type=code&redirect_uri=https://example.com/callback&scope=user-read-playback-state%20user-modify-playback-state%20user-read-currently-playing

**Comment by epicbenz** [comment]
> Try this url for the Auth Code and use his Client ID. In the url you provided above, you were missing a key part of the url:
> 
> https://accounts.spotify.com/authorize?client_id=abcd1234&response_type=code&redirect_uri=https://example.com/callback&scope=user-read-playback-state%20user-modify-playback-state%20user-read-currently-playing

Still the same error :/

I think I have found a difference, my Spotify Account is an old one, his is new and he is in his 30 days trial, and only after that he is a premium user? Maybe it's because of that.

Edit: Command back, I have tried the Spotify Account from my sister (also not in 30 days trial) and get the same error. Don't understand why my account works but their one not.  


Can I give u the access to the account and u try it out? 
Somewhere must be my mistake.

**Comment by joeyhage** [comment]
Is your sister’s account premium? I believe premium is required.

**Comment by joeyhage** [comment]
Another idea: when you ask for the client id from your friend/sister, are they also giving you the auth code? Because whoever the Client ID belongs to must be logged into Spotify in the same browser as the browser used for the Auth Code.

**Comment by epicbenz** [comment]
> Is your sister’s account premium? I believe premium is required.

Yes it's a Premium Family Account. The account of my friend is a Premium Single Account.

**Comment by epicbenz** [comment]
> Another idea: when you ask for the client id from your friend/sister, are they also giving you the auth code? Because whoever the Client ID belongs to must be logged into Spotify in the same browser as the browser used for the Auth Code.

Good point, but yes, I have all Loggins from them and have logged into their account on Spotify.com also. 

When I enter the address "https://accounts.spotify.com/authorize?client_id={clientId}&response_type=code&redirect_uri=https://example.com/callback&scope={scopes}" with their values I also have to agree some stuff and at the bottom there stands their account (is that you?) like it should.


Can I give u the access to the account and u try it out?
Somewhere must be my mistake.

**Comment by joeyhage** [comment]
Before we try that, we know your developer app is working. In the Spotify developer console, you can add more users to your developer app settings (up to 25). 

![IMG_4809](https://github.com/joeyhage/homebridge-spotify-speaker/assets/3220319/824264da-5eaa-4523-86c1-dd1a90cd37be)

can you try adding either your sister or friend, then get a new auth code - I would try in a private/incognito browser window to be safe, and then use the same client ID and client secret you have in your homebridge with their auth code. Let me know if that works!

**Comment by epicbenz** [comment]
> Before we try that, we know your developer app is working. In the Spotify developer console, you can add more users to your developer app (up to 25).
> 
> 
> can you try adding either your sister or friend, then get a new auth code - I would try in a private/incognito browser window to be safe, and then use the same client ID and client secret you have in your homebridge with their auth code. Let me know if that works!


You mean i should add the Spotify Email Accounts to my Developer Account under User Management?
Then generate a new AuthCode with my ClientID? Will my previous AuthCode from my Account than invalid?


I added my sister with name and her Spotify Email Adress, but in the settings I still have my ClientID. Is that correct? And now I should generate a new AuthCode with "https://accounts.spotify.com/authorize(...)" with my ClientID? Or what did you mean with their auth code?

I am a little worried that I crash my Account also :D

**Comment by epicbenz** [comment]
So I have added my sisters account to my Developer Site (from my account that worked)

<img width="1408" alt="Bildschirmfoto 2023-08-11 um 23 44 55" src="https://github.com/joeyhage/homebridge-spotify-speaker/assets/141502978/107bd8c1-177d-4306-925a-1dbd0d9bc279">

Then I went In a new private window on my browser to:
https://accounts.spotify.com/authorize?client_id={My Client ID from the Account that worked}&response_type=code&redirect_uri=https://example.com/callback&scope={scopes}
(with the scopes of your tutorial :D)

Then I have paste the AuthCode that I got into my Homebride Config (with still my Client ID and Secret ID). 

This worked. Do it help to find the error?

**Comment by joeyhage** [comment]
Great! So their must be a problem with the other developer apps. It will not cause any issues if you, your sister, and your friend all use the same developer app (Client ID + Secret) since you can add up to 25 users. 

I would suggest doing that then. As long as everyone has their own Auth Code from their own Spotify account, then they can only see their own songs, playlists, etc.

**Comment by joeyhage** [comment]
I’m still not sure why you were having issues with the other developer apps but we have another solution that will work just the same

**Comment by epicbenz** [comment]
> Great! So their must be a problem with the other developer apps. It will not cause any issues if you, your sister, and your friend all use the same developer app (Client ID + Secret) since you can add up to 25 users.
> 
> I would suggest doing that then. As long as everyone has their own Auth Code from their own Spotify account, then they can only see their own songs, playlists, etc.


Unfortunately, it still does not work. If I now link my sister's Spotify account on the Echo Dot, I get the following log in my Homebridge:

```
[8/12/2023, 9:26:07 AM] [HomebridgeSpotifySpeaker] spotifyDeviceName 'Andreas' 2. Echo Dot' did not match any Spotify devices. spotifyDeviceName is case sensitive.
```

It only worked with my Spotify Account on the Echo Dot, but then they are using my Account and I can't use the account in the same time. 

Don't really know what to do or where my error could be (tried it in a new browser but still get the same error like at the top).

My friends Spotify Account is only for this use of case, so can I give u his login and you try it out?

**Comment by joeyhage** [comment]
I think we are saying different things.

Client ID and Secret do not determine which Spotify account is used. Auth Code is a replacement for username/password but more secure and required by Spotify.

An Echo dot can only be used for one Spotify account, not multiple. So any Echo dots owned by you should be logged into your Spotify account only. 

In order for this to work for all three of you, you need:

1. Each person needs their own homebridge server
2. Each person needs their own Echo device logged into their own Spotify account
3. Each person needs their own auth code that was obtained in a private/incognito window signed into their Spotify account.
4. Client ID and Client Secret can be the same as long as the users are in the Spotify developer user management


Giving me the username/password won’t be helpful since I cannot access the homebridge server or the Echo dot.

The error message you shared means it logged in successfully - so what devices are listed in the homebridge logs now? If you see your echo dots, that means the Auth Code came from your Spotify account.

**Comment by epicbenz** [comment]
> I think we are saying different things.
> 
> Client ID and Secret do not determine which Spotify account is used. Auth Code is a replacement for username/password but more secure and required by Spotify.
> 
> An Echo dot can only be used for one Spotify account, not multiple. So any Echo dots owned by you should be logged into your Spotify account only.
> 
> In order for this to work for all three of you, you need:
> 
> 1. Each person needs their own homebridge server
> 2. Each person needs their own Echo device logged into their own Spotify account
> 3. Each person needs their own auth code that was obtained in a private/incognito window signed into their Spotify account.
> 4. Client ID and Client Secret can be the same as long as the users are in the Spotify developer user management
> 
> Giving me the username/password won’t be helpful since I cannot access the homebridge server or the Echo dot.
> 
> The error message you shared means it logged in successfully - so what devices are listed in the homebridge logs now? If you see your echo dots, that means the Auth Code came from your Spotify account.

Yes my English is not the best :D 

So again for clarification: 
I'm trying to set up on my Homebridge (even deleted and reinstalled the plugin once) with all the variables from my friend's account (Client ID, Client Secret, generated AuthCode) to test it and then if everything works then set it up on his Homebridge (not at the same time of course). 

On the Echo Dot I also logged in with his Spotify account with the Spotify Skill. 

So I wanted to set up the whole thing from scratch, so to speak, with a different account. 


In the Homebridge logs are unfortunately no devices, so far it goes unfortunately not at all. I only get the hint:


```
We could not fetch the Spotify tokens nor authenticate using the code grant flow.
```
And:
```
Could not authorize Spotify
```


I really don't understand it, I really did everything the same as with the first account. It can't be the Homebridge (but I deleted the plugin to be on the safe side and set it up again), unfortunately without success. 


I have an old Raspberry Pi, I will setup a new Homebridge on that and will try with the account data of my friend. Don't think that this will make a difference but I try it.

**Comment by epicbenz** [comment]
Okay, don't know why but on the completely new Homebridge on the Raspberry Pi it worked! 
Can it be, that its not enough to just delete the Plugin to change the Account? 
Anyways, it works :D 

Thank you very much!

**Comment by joeyhage** [comment]
@epicbenz thanks for your detailed explanation of what you were attempting to do! That was very helpful. There is one file the plug-in stores information in that would not be deleted when the plug-in is uninstalled. 

I don’t believe the plug-in can delete that file automatically when the plug-in is uninstalled but I will research into alternative options. 

Glad you found a workaround for now!

**Comment by cjdrawdy** [comment]
@joeyhage What is the path of the file that is not automatically deleted when the plug in is uninstalled? I uninstalled while troubleshooting another issue and am having auth issues now that I've re-installed.

**Comment by joeyhage** [comment]
@cjdrawdy the file is `persist/.homebridge-alexa-smarthome` located in the homebridge installation directory

**Comment by cjdrawdy** [comment]
@joeyhage the Alexa plugin holds credentials for spotify?

**Comment by cjdrawdy** [comment]
@joeyhage - I'm not seeing this path. I'm using homebridge-ui-x.
